runtime/signals.c

```diff
@@ -147,6 +148,9 @@ CAMLexport void caml_enter_blocking_section(void)
     caml_enter_blocking_section_hook ();
     /* Check again for pending signals.
        If none, done; otherwise, try again */
+    // FIXME: does this become very slow if a signal is recorded but
+    // is masked for everybody in capacity of running signals at this
+    // point?
```

Comment:
This is a correctness concern for the code currently in trunk. We test whether any signal is pending, not just ours. It could be that a signal not for us (masked) is recorded and its processing by the target thread takes time. In this situation running `caml_process_pending_signals_exn` is very expensive because it is guarded by the same test (in addition it contains system calls). Since running it does not end with all signal handlers being called, it is called again in a loop. Let me know if my concerns are justified. See 16th commit about "Do not wait that other threads are processes their signals" for a proposed fix. (cc @xavierleroy)